Define IConfidentialToken base interface

awrichar opened the pull request:
This defines a basic confidential UTXO token interface that could potentially be implemented by both Noto and Zeto (and any other similar tokens in the future).

awrichar committed:
Define IConfidentialToken base interface

This defines a basic confidential UTXO token interface that could potentially be
implemented by both Noto and Zeto (and any other similar tokens in the future).

Signed-off-by: Andrew Richardson <andrew.richardson@kaleido.io>

awrichar committed:
Merge branch 'main' into confidential-token

peterbroadhurst commented on solidity/contracts/domains/interfaces/IConfidentialToken.sol:
```diff
@@ -0,0 +1,55 @@
+     */
+    function transfer(
+        bytes32 txId,
+        bytes32[] calldata inputs,
```
What about "read" and "info" states that are needed for other data not mutated.

Few examples:
- Allow-lists for ZKPs ("read")
- Account states for private EVM execution that are not modified ("read")
- Transaction input payload reference ("info")
- Transaction manifest for establishing subsets of states visible to each party ("info")

annamcallister committed:
Merge branch 'main' into confidential-token

awrichar committed:
Merge branch 'main' into confidential-token

awrichar committed:
Replace "UTXO" terminology with just "states"

Signed-off-by: Andrew Richardson <andrew.richardson@kaleido.io>

awrichar committed:
Extract generic ILockableCapability interface

This interface is not specific to C-UTXO tokens, but can describe how to
lock and unlock any arbitrary operation.

Signed-off-by: Andrew Richardson <andrew.richardson@kaleido.io>

awrichar commented on solidity/contracts/domains/interfaces/IConfidentialTokenLockable.sol:
```diff
@@ -0,0 +1,148 @@
+        bytes32[] outputs;
+        // Any additional transaction data (opaque to the blockchain)
+        bytes data;
+    }
```
Correct - it's purely for gas efficiency to avoid storage cost.

awrichar commented:
Superseded by https://github.com/LFDT-Paladin/paladin/pull/978
